Geochemistry: time dependence

WilkAndy opened the pull request:
TODO:
 - exception tests for all new MOOSE classes
 - simple tests for relevant MOOSE classes in which a complicated solve is unnecessary
 - thorough tests that all the parameters supplied to Actions are making their way through to the Objects
 - documentation in some header files and ClassDescriptions
 - markdown documentation for all MOOSE classes
 - when temperature-dependence is implemented, check that the results are precisely correct

Refs #15330

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

WilkAndy committed:
Geochemistry: time dependence

TODO:
 - exception tests for all new MOOSE classes
 - simple tests for relevant MOOSE classes in which a complicated solve is unnecessary
 - thorough tests that all the parameters supplied to Actions are making their way through to the Objects
 - documentation in some header files and ClassDescriptions
 - markdown documentation for all MOOSE classes
 - when temperature-dependence is implemented, check that the results are precisely correct

Refs #15330

WilkAndy committed:
tests and doco

Fixes #15330

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/541801/) on b0c9154 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/15409/site)

This comment will be updated on new commits.

loganharbour commented:
@WilkAndy - We've got conferences and a workshop going on this week. I may get to this tomorrow but I may not be able to finish until Thursday, so no promises :-)

WilkAndy commented:
OK, thanks @loganharbour

WilkAndy committed:
kinetics: some core utilities

Although i didn't initially think to include kinetics in this PR, they are 'time-dependent', so they are not inappropriate here.  Note there is no actual functionality available to users yet - this is just utils stuff.

Refs #15330
Refs #15478

WilkAndy commented:
@aeslaughter , i'm wondering if you can help with the documentation error here.  At https://civet.inl.gov/job/545660/ you can see the error:
```
====== Testing MooseDocs in /opt/civet/build_0/moose/modules/geochemistry ======
MooseDocs.Executioner (MainProcess): Executing extension init() methods...
MooseDocs.Executioner (MainProcess): Executing extension init() methods complete [7.009506225585938e-05 sec.]
MooseDocs.Executioner (MainProcess): Executing extension initPage() methods...
MooseDocs.Executioner (MainProcess): Executing extension initPage() methods complete [0.1983950138092041 sec.]
MooseDocs.build (MainProcess): Cleaning destination /opt/civet/build_0/moose/site
MooseDocs.extensions.appsyntax (MainProcess): Reading MOOSE application syntax...
MooseDocs.extensions.appsyntax (MainProcess): MOOSE application syntax complete [4.874826192855835 sec.]
MooseDocs.extensions.appsyntax (MainProcess): Building MOOSE class database...
MooseDocs.extensions.appsyntax (MainProcess): MOOSE class database complete [3.5189337730407715 sec]
MooseDocs.Translator (MainProcess): Executing preExecute methods...
MooseDocs.Translator (MainProcess): Executing preExecute methods complete [0.8365552425384521 sec.]
MooseDocs.Executioner (MainProcess): Translating using 16 threads...
MooseDocs.base.renderers (Process-11): 
RENDER ERROR: The shortcut link key 'mass' was not located in the list of shortcuts.
/opt/civet/build_0/moose/framework/doc/content/syntax/index.md:1
 \u250c\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2510
1\u2502examples                                                                                         \u2502
 \u2514\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2518
Traceback (most recent call last):
  File "/opt/civet/build_0/moose/python/MooseDocs/base/renderers.py", line 85, in render
    el = func(parent, token, page) if func else parent
  File "/opt/civet/build_0/moose/python/MooseDocs/extensions/core.py", line 433, in createHTML
    node = self.getShortcut(token)
  File "/opt/civet/build_0/moose/python/MooseDocs/extensions/core.py", line 467, in getShortcut
    raise exceptions.MooseDocsException(msg, key)
MooseDocs.common.exceptions.MooseDocsException: The shortcut link key 'mass' was not located in the list of shortcuts.
```
What does this mean?

loganharbour commented:
@WilkAndy I haven't touched this after we discussed conflicts in the other PR.

Are we actually good to go? Other than the doco question above.

WilkAndy commented:
@loganharbour - yep, it's ready for review.  i'll be adding more if we're not careful!

WilkAndy commented:
@aeslaughter - yea, must be some kind of doco error, but what does the `shortcut link key 'mass' was not located in the list of shortcuts.
/opt/civet/build_0/moose/framework/doc/content/syntax/index.md:1` mean?  Does the information provided above give you any hints as to where the error is?  I looked for a link to 'mass' in geochemistry and couldn't find anything.

aeslaughter commented on modules/geochemistry/src/userobjects/GeochemistryKineticRate.C:
```diff
@@ -0,0 +1,95 @@
+      "a single kinetic_species: the sum the individual rates defines the overall rate.  "
+      "GeochemistryKineticRate simply specifies the algebraic form for a kinetic rate: to actually "
+      "use it in a calculation, you must use it in the GeochemicalModelDefinition.  The rate is "
+      "intrinsic_rate_constant * area_quantity * [mass of kinetic_species in grams] * "
```
I think this is the error, the system is trying to lookup a markdown shortcut with the content in []. Either change this to () or escape the brackets with \[ \].

aeslaughter replied:
```suggestion
      "intrinsic_rate_constant * area_quantity * \[mass of kinetic_species in grams\] * "
```

WilkAndy replied:
Fantastic, thank you @aeslaughter

WilkAndy commented:
Don't get too depressed, @loganharbour , a fair amount of the 16k lines is comprehensive unit testing.

WilkAndy committed:
kinetic additions to geochemistry
ability to set ramp_max_ionic_strength during timestepping
remove uncharged kinetic contributions from stoichiometric ionic strength calculation - they should not dissociate since they are kinetic species
also kinetic-related changes to "flush" mode, which i believe is now like Bethke definition
kinetic and flush simulation examples added and documented

Refs #15330
Fixes #15489

loganharbour commented on modules/geochemistry/include/utils/EquilibriumGeochemicalSystem.h:
```diff
@@ -357,6 +363,218 @@ class EquilibriumGeochemicalSystem
+   * preceeding paragraphs (setting to zero for equilibrium minerals and gases, and constraints
+   * overriding the values provided)!
+   */
+  void setSolventMassAndFreeMolalityAndMineralMolesAndSurfacePots(
```
Talk about a handful

WilkAndy replied:
Yep, super long.  Evidently i wanted to be really clear what was going on in this method.

loganharbour commented:
I have a big question here: why not scalar aux variables or postprocessors? Why aux variables?

WilkAndy commented:
> I have a big question here: why not scalar aux variables or postprocessors? Why aux variables?

Yea, good question.   Certainly for this spatially-independent stuff scalar/postprocessors would be much more suitable.  I'm looking towards the next chunk of work, which will be spatially-dependent, when AuxVars will be necessary.  Maybe this will turn out to be a silly idea.... time will tell.

loganharbour commented:
> > I have a big question here: why not scalar aux variables or postprocessors? Why aux variables?
> 
> Yea, good question. Certainly for this spatially-independent stuff scalar/postprocessors would be much more suitable. I'm looking towards the next chunk of work, which will be spatially-dependent, when AuxVars will be necessary. Maybe this will turn out to be a silly idea.... time will tell.

Hmm... fair enough. I'll get back to it, then.

loganharbour commented:
SO... what's the status? ready again?

WilkAndy commented:
Always ready, @loganharbour .   Actually, today i want to write a few more unit tests :-)

a

WilkAndy commented:
Wow, so this is another colossal addition to `geochemistry`: 19000 lines!  In conclusion, both time-dependent reactions (eg, adding reactants to a test-tube) and kinetically-controlled reactions can now be modelled.

@cpgr and @rpodgorney , this completes a major milestone of the `geochemistry` module.  We can now do virtually every type of reaction process that other geochemistry solvers can do.  The module now contains the entire reaction-based functionality described at https://mooseframework.org/modules/geochemistry/ including all the tests and examples at https://mooseframework.org/modules/geochemistry/tests_and_examples/geochemistry_tests_and_examples.html  .   There are 300 tests in the test suite - 200 unit tests that are mostly boring things, and 100 MOOSE tests that are also mostly boring things, but a few realistic examples are included.

There are two remaining coding milestones (not in this PR, thank goodness!)

- thoroughly check the benchmark examples (comparisons with the Bethke book).  Currently i can't do this as the temperature-dependent equilibrium constants and activity coefficients haven't been included.  Including them is trivial, but i imagine that thorough benchmarking will uncover some problems, or at least will take some weeks.

- include transport.  Even though i've always kept this in my mind, i'm sure it will require a fair refactoring of the current code.  And a fair amount of lines of code, sorry @loganharbour .

rpodgorney commented:
Holy Dooley @WilkAndy, that is colossal!  Well done, i just don't know how get get so much code written so quickly.  @permcody and @loganharbour -- I know checking all of this will be a big task for you guys, let me know how i can help or incentivize. Can't wait to get going with this capability.

WilkAndy committed:
unit/database/moose_testdb.json updated to have floats instead of strings.  Refs #15145

WilkAndy committed:
temperature dependence added to the Integrrogator.  Fixes #15144

WilkAndy commented:
@loganharbour ,

- i rebased against `next` which was good since this PR would have caused all the geochem unit tests to fail - a bit of a puzzle for future developers (me!).  Shame that modules unit tests aren't being run by CIVET.

- i'm slowly adding temperature dependence to this commit as well as time-depependence.  I don't think this will impact your review much, since it's very few lines of code and is mostly lots of testing - benchmark comparisons with other codes.  It will take me quite a while to do all the tests once the C++ is done, so you'll have a while for your review

WilkAndy committed:
Ability to use a piecewise-linear interpolation instead of 4th-order or maier-kelly in EquilibriumConstantInterpolator.  Fixes #15577

WilkAndy committed:
Temperature-dependent activity coefficients.  Fixes #15593

GeochemicalDatabaseReader now has some const methods for getting temperature-related things, so that we can ensure that nothing modifies the database

GeochemicalDatabaseValidator now checks for "note" lines that come from interpolation of 500.0000 quantities

GeochemistryActivityCoefficientsDebyeHuckel now does full temperature interpolation of the DebyeHuckel parameters

database/moose_testdb.json now contains more DebyeHuckel parameters: makes testing easier
test/database databases specify fitting scheme

temperature-dependence has implied miscellaneous small changes to:
 - GeochemicalDatabaseReaderTest.C GeochemicalSolverTest.C GeochemicalSystemTest.C GeochemistryActivityCoefficientsDebyeHuckelTest.C
 - various gold files

WilkAndy committed:
temperature-dependent logK.  Fixes #15594

ModelGeochemicalDatabase struct in PertinentGeochemicalSystem now has a pointer to the original database that it was created with.  This means
- it does not need to hold the temperature points, and also:
- GeochemicalModelInterrogator does not need its _db anymore - can get that from _mgd

GeochemicalSystem - now does temperature-dependent logK, yippee!

Various re-goldings of tests needed

WilkAndy committed:
make unit tests more robust to re-orderings of redox info

skip some tests that fail on CIVET - these are temp-dep tests that need to be validated

WilkAndy committed:
Variety of small changes:

- GeochemicalModelDefinition now has a flag that allows a piecewise-linear fit to be used instead of whatever is specified in the database.  Refs #15577
- stoichiometric ionic strength can be set to chlorine molality.  Fixes #15601
- moles_per_kg_water changed to 55.5 instead of 55.51 for better comparisons with GWB.  Fixes #15603
- minor regolds of a few tests due to these changes

and

Test HCl fully documented and golded wrt GWB.  Refs #15293

WilkAndy committed:
Test seawater fully documented and golded wrt GWB.  Refs #15293

moosebuild commented:
Job [Create Mac Stack](https://civet.inl.gov/job/565390/) on f061423 : invalidated by @milljm

Initialize Conda Environment never finished successfully. Deleted environment. Trying again.

WilkAndy committed:
Test amazon fully documented and golded wrt GWB.  Refs #15293

WilkAndy committed:
e- removed from equilibrium species list and put into redox info.  Wow, bigger jost than i thought.  Fixes #15606

WilkAndy committed:
Test red_sea fully documented and golded wrt GWB.  Refs #15293

WilkAndy committed:
Test morro fully documented and golded wrt GWB.  Refs #15293

WilkAndy committed:
Test microbial fully documented and golded wrt GWB.  Refs #15293

WilkAndy committed:
Order Nernst potential output in GeochemistryConsoleOutput.  Fixes  #15620

WilkAndy committed:
removed superseded classes EquilibriumGeochemicalSolver and EquilibriumGeochemicalSystem.  They have been superseded by GeochemicalSolver and GeochemicalSystem.  Fixes #15619

WilkAndy committed:
Tests involving reaction interrogations fully documented and golded wrt GWB.  Refs #15293

WilkAndy committed:
Test gypsum fully documented and golded wrt GWB.  Refs #15293

WilkAndy committed:
Test sebkhat_el_melah fully documented and golded wrt GWB.  Refs #15293

WilkAndy committed:
Test selenate fully documented and golded wrt GWB.  Refs #15293

WilkAndy committed:
Test add_feldspar fully documented and golded wrt GWB.  Refs #15293

WilkAndy committed:
fixed factor of 8 in surface-potential calcs.  Fixes #15626

WilkAndy committed:
remove surface-sorption related things from ionic strength calcs.  Fixes #15627

WilkAndy committed:
removed surface-sorption species from reported solution mass in GeochemistryConsoleOutput.  Fixes #15628

WilkAndy committed:
Test ferric_hydroxide  fully documented and golded wrt GWB.  Refs #15293

WilkAndy committed:
Test changing_pH_ferric_hydroxide fully documented and golded wrt GWB.  Refs #15293

WilkAndy committed:
flag that allows removal of secondary species with extrapolated logK values.  Fixes #15629

WilkAndy committed:
Test cooling fully documented and golded wrt GWB.  Refs #15293

WilkAndy committed:
Tests kinetic_albite and quartz_dissolution fully documented and golded wrt GWB.  Refs #15293

WilkAndy committed:
Test quartz_deposition fully documented and golded wrt GWB.  Refs #15293

WilkAndy committed:
Tests dissolution_pyrite_1 and dissolution_pyrite_2 fully documented and golded wrt GWB.  Refs #15293

WilkAndy committed:
Test changing_fugacity_calcite fully documented and golded wrt GWB.  Refs #15293

WilkAndy committed:
Test mixing fully documented and golded wrt GWB.  Refs #15293

WilkAndy committed:
Doco describing the differences between geochemistry and GWB.  Refs #15641

WilkAndy committed:
Test calcite_dumping fully documented and golded wrt GWB.  Refs #15293

WilkAndy committed:
Ability to record moles of minerals dumped in dump and flow-through modes.  Fixes #15643

WilkAndy committed:
Test seawater_evaporation fully documented and golded wrt GWB.  Refs #15293

WilkAndy committed:
Test flushing documented and golded wrt GWB.  Refs #15293

WilkAndy committed:
doco reorganized.  Fixes #15652

WilkAndy committed:
flushing test re-done.  See if it passes CIVET.  Refs #15653

WilkAndy commented:
Hi @loganharbour .    This is ready for review.    I'm sorry that i've been adding and adding and adding.  But i'm on a really tight time frame and can't stop developing.  It's all like a big tower - i just have to keep building and building on top, and can't build something else.

WilkAndy commented:
Hi @cpgr and @rpodgorney .   I'm quite pleased because i've documented all the benchmark tests, and they all agree nicely with the Geochemists Workbench software.  So i'm confident that we can say "we have a MOOSE geochemical solver" now.   Next step - transport.

WilkAndy committed:
Rewording in gwb_diff.md    Fixes #15641

rpodgorney commented:
Awesome work Andy!  I’ve been watching all the commits you’ve been making and just amazed at your level of productivity!!!

Thanks a ton.

Robert Podgorney, PhD, PG
Lead, Geothermal and Fossil Energy
Energy and Environment Division
Idaho National Laboratory

Office: 208-526-1524<tel:208-526-1524>
Cell: 208-520-9361<tel:208-520-9361>
email: robert.podgorney@inl.gov<mailto:robert.podgorney@inl.gov>

-sent from my mobile device, please excuse any typos or autocorrect errors

Statement of Confidentiality
The contents of this e-mail message and any attachments are confidential and are intended solely for addressee. The information may also be legally privileged. This transmission is sent in trust, for the sole purpose of delivery to the intended recipient. If you have received this transmission in error, any use, reproduction or dissemination of this transmission is strictly prohibited. If you are not the intended recipient, please immediately notify the sender by reply e-mail or phone and delete this message and its attachments, if any.


On Jul 22, 2020, at 22:44, Andy Wilkins <notifications@github.com> wrote:

﻿

Hi @cpgr<https://gcc01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fcpgr&data=02%7C01%7Crobert.podgorney%40inl.gov%7Cb50beff3912a40020c4d08d82ec30b0d%7C4cf464b7869a42368da2a98566485554%7C0%7C0%7C637310762531583856&sdata=%2BTzmG08G2MEOBacTHnVqTWWxXqVsU1B%2F2eNoC0vxCnU%3D&reserved=0> and @rpodgorney<https://gcc01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Frpodgorney&data=02%7C01%7Crobert.podgorney%40inl.gov%7Cb50beff3912a40020c4d08d82ec30b0d%7C4cf464b7869a42368da2a98566485554%7C0%7C0%7C637310762531593814&sdata=%2FI1%2BzUvTNpWkzOTlDvFTTu4pL8nol8l6wTbb5gklxj4%3D&reserved=0> . I'm quite pleased because i've documented all the benchmark tests, and they all agree nicely with the Geochemists Workbench software. So i'm confident that we can say "we have a MOOSE geochemical solver" now. Next step - transport.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://gcc01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fidaholab%2Fmoose%2Fpull%2F15409%23issuecomment-662815752&data=02%7C01%7Crobert.podgorney%40inl.gov%7Cb50beff3912a40020c4d08d82ec30b0d%7C4cf464b7869a42368da2a98566485554%7C0%7C0%7C637310762531593814&sdata=jFS%2F7otuN6j%2FjrJY8WZ0EFG3OVYKHFj7ubemIc8oBJQ%3D&reserved=0>, or unsubscribe<https://gcc01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FACWGDT5ZMJCOZYKDTCQIXQLR465ZVANCNFSM4NTE3NEA&data=02%7C01%7Crobert.podgorney%40inl.gov%7Cb50beff3912a40020c4d08d82ec30b0d%7C4cf464b7869a42368da2a98566485554%7C0%7C0%7C637310762531593814&sdata=v9%2Fq7nfJGRH9g%2FNOshn%2BGFoMH8qSv9S0aVvTaS9EsO4%3D&reserved=0>.

WilkAndy commented:
Bugging @loganharbour

moosebuild commented:
Job [Mac Test](https://civet.inl.gov/job/571638/) on 3b56a6e : invalidated by @loganharbour

unrelated failure

loganharbour commented:
On it. Are you gonna have someone review the doco?

WilkAndy commented:
Probably @cpgr can look at the doco.  Actually, it hasn't changed much.  I basically did a small reorganisation, and then added a lot of links to input files, and figures showing agreement with benchmarks.

cpgr commented:
I'll take a look this weekend

WilkAndy commented:
Thanks Chris, but you probably can wait till next week :-)

WilkAndy commented:
Bugging @loganharbour

loganharbour reviewed (commented):
I am still working on this, by the way :-)

I've gotten through a little more than half of it and haven't looked at the doco yet. I will leave this comment though, as this is the most glaring thing to me right now.

WilkAndy commented on modules/geochemistry/include/userobjects/GeochemistryReactorBase.h:
```diff
@@ -0,0 +1,108 @@
+  /**
+   * @return a reference to the equilibrium geochemical system at the given node
+   * @param node_id the ID of the node
+   */
+  virtual const GeochemicalSystem & getGeochemicalSystem(unsigned node_id) const = 0;
+
+  /**
+   * @return a reference to the most recent solver output (containing iteration info, swap info,
+   * residuals, etc)
+   * @param point the point of interest
+   */
+  virtual const std::stringstream & getSolverOutput(const Point & point) const = 0;
+
+  /**
+   * @return the total number of iterations used by the most recent solve at the point
+   * @param point the point of interest
+   */
+  virtual unsigned getSolverIterations(const Point & point) const = 0;
+
+  /**
+   * @return the L1norm of the residual at the end of the most recent solve at the point
+   * @param point the point of interest
+   */
+  virtual Real getSolverResidual(const Point & point) const = 0;
+
+  /**
+   * @return the mole additions (the first num_basis of these are additions to the bulk composition
+   * of the equilibrium system, the last num_kin of these are -dt*kinetic_reaction_rate, ie
+   * dt*dissolution_rate)
+   * @param node_id the node ID
+   */
+  virtual const DenseVector<Real> & getMoleAdditions(unsigned node_id) const = 0;
+
+  /**
+   * @return the mole additions (the first num_basis of these are additions to the bulk composition
+   * of the equilibrium system, the last num_kin of these are -dt*kinetic_reaction_rate, ie
+   * dt*dissolution_rate)
+   * @param point the point of interest
+   */
+  virtual const DenseVector<Real> & getMoleAdditions(const Point & point) const = 0;
```
There is the getMoleAdditions(unsigned node_id) for more serious usage like filling AuxVariables.  This getMoleAdditions(const Point & point) is just for the ConsoleOutput object which writes the result of a simulation at the given point to std::out, rather like a PointValue postprocessor.  I'm happy with there being roundoff or other effects in that output.

loganharbour replied:
Got it 👍🏼 just looked into this.

cpgr commented:
I've started too but there is a lot to go through!

loganharbour commented on modules/geochemistry/include/utils/GeochemicalSystem.h:
```diff
@@ -0,0 +1,934 @@
+  kin_initial_moles.  All kinetic species must be provided with an initial condition
+   * @param kin_initial_moles values of the initial mole number for the kinetic species
+   */
+  GeochemicalSystem(ModelGeochemicalDatabase & mgd,
+                    GeochemistryActivityCoefficients & gac,
+                    const GeochemistryIonicStrength & is,
+                    GeochemistrySpeciesSwapper & swapper,
+                    const std::vector<std::string> & swap_out_of_basis,
+                    const std::vector<std::string> & swap_into_basis,
+                    const std::string & charge_balance_species,
+                    const std::vector<std::string> & constrained_species,
+                    const std::vector<Real> & constraint_value,
+                    const MultiMooseEnum & constraint_meaning,
+                    Real initial_temperature,
+                    unsigned iters_to_make_consistent,
+                    Real min_initial_molality,
+                    const std::vector<std::string> & kin_name,
+                    const std::vector<Real> & kin_initial_moles);
+
+  GeochemicalSystem(ModelGeochemicalDatabase & mgd,
+                    GeochemistryActivityCoefficients & gac,
+                    const GeochemistryIonicStrength & is,
+                    GeochemistrySpeciesSwapper & swapper,
+                    const std::vector<std::string> & swap_out_of_basis,
+                    const std::vector<std::string> & swap_into_basis,
+                    const std::string & charge_balance_species,
+                    const std::vector<std::string> & constrained_species,
+                    const std::vector<Real> & constraint_value,
+                    const std::vector<ConstraintMeaningEnum> & cc,
+                    Real initial_temperature,
+                    unsigned iters_to_make_consistent,
+                    Real min_initial_molality,
+                    const std::vector<std::string> & kin_name,
+                    const std::vector<Real> & kin_initial_moles);
```
You've got a _ton_ of code duplication here in constructors but I can't seem to find a way to make it much better!

loganharbour commented on modules/geochemistry/src/actions/AddGeochemistrySolverAction.C:
```diff
@@ -0,0 +1,344 @@
+  params.addRequiredParam<UserObjectName>("model_definition",
+                                          "The name of the GeochemicalModelDefinition user object "
+                                          "(you must create this UserObject yourself)");
+  params.addParam<UserObjectName>(
```
FYI - as you seem to be growing these large lists of params in your actions that are duplicates of the objects they create, I have a suggestion. Take, for example, `GeochemistryConsoleOutput`.

You can create an additional param static function, say `static InputParameters GeochemistryConsoleOutput::sharedParams()`. Here is where you would put the params that you want to appear both in the console output object and the action that creates it. Then, you would have
```
InputParameters
GeochemistryConsoleOutput::validParams()
{
  InputParameters params = Output::validParams();
  params += GeochemistryConsoleOutput::sharedParams();
}
```
This would then allow you to do something similar here, in which you can just add all of the shared params from another object without having to duplicate them.

WilkAndy replied:
Oh cool, i didn't know about that.  Thanks - i'll address that today

WilkAndy replied:
Done the sharedParams idea, thanks @loganharbour - got rid of 262 lines of duplicated code!

loganharbour replied:
Awesome! Looks great!

WilkAndy replied:
Thanks @loganharbour  :-)

loganharbour commented on modules/geochemistry/src/actions/AddTimeDependentReactionSolverAction.C:
```diff
@@ -0,0 +1,218 @@
+                    "add_geochemistry_reactor");
+
+InputParameters
+AddTimeDependentReactionSolverAction::validParams()
```
Same thing here about crazy code duplication for input params!

WilkAndy replied:
Yep, done that, thanks

loganharbour commented on modules/geochemistry/src/outputs/GeochemistryConsoleOutput.C:
```diff
@@ -0,0 +1,318 @@
+}
+
+void
+GeochemistryConsoleOutput::output(const ExecFlagType & type)
```
Not necessary to attack in this PR,  but we do have some great table output utilities that might prettify your output some :)

WilkAndy replied:
Oh, that might be nice.  Let's leave it for now, probably, but can you show me an example?

loganharbour reviewed (commented):
This is actually looking pretty great: great error checking, const where you need it, and lots of useful comments for someone who is separated from the physics. I don't have any glaring concerns that could be addressed other than maybe removing some of the duplication for params. 

@cpgr have you had a chance to look through this some?

WilkAndy committed:
sharedParams used in GeochemicalModelInterrogator, GeochemistryConsoleOutput, GeochemistryReactorBase and GeochemistryTimeDependentReactor so they can be used in the various Actions to reduce code duplication.  Refs #15478

loganharbour commented:
If this wasn't clear - I'm good with this except for an extra check over the documentation if possible.

WilkAndy commented:
> If this wasn't clear - I'm good with this except for an extra check over the documentation if possible.

Thanks @loganharbour .    Over to you, @cpgr

cpgr commented:
Mea culpa - I haven't finished yet. Promise to have it done by Monday.

Have a great weekend guys

cpgr commented on modules/geochemistry/doc/content/modules/geochemistry/tests_and_examples/changing_pH_iron.md:
```diff
@@ -2,45 +2,34 @@
+
+!listing modules/geochemistry/test/tests/time_dependent_reactions/changing_pH_ferric_hydroxide.i block=Postprocessors
+
+[!cite](bethke_2007) presents results in Figures 14.8 and 14.9 (bold line only).  The results are faithfully reproduced by the `geochemistry` module as shown in the figures below.
```
What is bold line only? Are there multiple lines in the Bethke figures?

cpgr reviewed (commented):
The documentation is so comprehensive, especially the comparisons with GWB. Well done Andy!

This shouldn't hold up this PR, but I think that we should think up a way to use the default database without having to specify it in the input file, or at least not have to specify the path.

And what about if there was a `GWB_mode` that could be set in an action so that it automatically set all of the input parameters needed to mimic GWB exactly? I don't even know if that would be possible?

Anyway, they are for future work. Well done again Andy on this mountain of work

WilkAndy commented:
> The documentation is so comprehensive, especially the comparisons with GWB. Well done Andy!
> 
> This shouldn't hold up this PR, but I think that we should think up a way to use the default database without having to specify it in the input file, or at least not have to specify the path.
> 
> And what about if there was a `GWB_mode` that could be set in an action so that it automatically set all of the input parameters needed to mimic GWB exactly? I don't even know if that would be possible?
> 
> Anyway, they are for future work. Well done again Andy on this mountain of work

Re the path: if this were python we could looking for the path of the executable and use path/database/blah.json.  Not sure how easy that would be in C++, when including problems with symbolic links, or windows-style paths, etc.  Not sure whether users with their own apps would find it annoying or useful.  I think this is something you could think about in your free time, haha!

Re the GWB mode: i'm not super confident of this, as i'm not really sure i know exactly how GWB works.  i discovered the various things like stoichiometry ionic strength using Cl-, and the temperature interpolations not being active at special temnperatures, rather by accident.  There could be a lot of other things i haven't yet encountered.  Something that really bugs me personally is using 55.5 moles of water per kg, and ln(10)=2.303, which is what GWB uses.  If we had your flag then we could use those approximations only when the flag is on, which would make me much happier.  But as you say, let's leave this for another day....

a

WilkAndy committed:
addressing @cpgr comments

loganharbour commented:
> Re the path: if this were python we could looking for the path of the executable and use path/database/blah.json. Not sure how easy that would be in C++, when including problems with symbolic links, or windows-style paths, etc. Not sure whether users with their own apps would find it annoying or useful. I think this is something you could think about in your free time, haha!

We could probably do something simple for when `$MOOSE_DIR` is available or can be inferred. If so, no file path is required. Otherwise - error and ask for a path.

loganharbour reviewed (approved):
Cheers to another massive contribution 🍻

WilkAndy commented:
This must have taken you ages to read, thank you so much @loganharbour
